### PR TITLE
Roles with nested Permissions

### PR DESCRIPTION
**Pull request by Redjik**
UserProxy added to clarify Rule api
Roles with nested Permissions
Route parsing for permissions

### CODE REVIEW

File: api/app/Http/Auth/ManipulateWithOwn.php [hunk 1]
```diff
@@ -10,23 +10,20 @@
...
      * @return bool a value indicating whether the rule permits the auth item it is associated with.
      */
-    public function execute(callable $userResolver, $params)
+    public function execute(UserProxy $userProxy, $params)
     {
-        /** @var Authenticatable $user */
-        $user = $userResolver();
-
-        return isset($params['model']) ? $params['model']->user_id == $user->getAuthIdentifier() : false;
+        return isset($params['model']) ? $params['model']->user_id == $userProxy->resolveUser()->getAuthIdentifier() : false;
```
Comment (zakhenry) [inline api/app/Http/Auth/ManipulateWithOwn.php:27]: Yea this is better for typehinting and code completion.

**Comment by zakhenry** [comment]
@Redjik the structure is perfect, however can I get you to refactor it so that the permissions are defined as a relationship, so that the permissions can be retrieved with the user (using with-nested). eg `GET /users/{userId} With-Nested: roles.permissions`

I realise that we don't have a table for it, but we have implemented virtual relationships before. See: `Spira\Model\Model\VirtualRelationInterface` - there is an implementation of this for articleComments

File: api/app/Http/Auth/ManipulateWithOwn.php (hunk 1 above)
Comment (Redjik) [inline api/app/Http/Auth/ManipulateWithOwn.php:27]: Not only, this way we define contract in Rule object, not just agree between ourselves that we should `$user = $userResolver();` to resolve the user.
Much easier for the newcomers.

File: api/app/Models/Permission.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace App\Models;
+
+use Spira\Model\Model\BaseModel;
+
+class Permission extends BaseModel
```
Comment (zakhenry) [inline api/app/Models/Permission.php:15]: This should probably extend `VirtualModel` as it has already disabled the methods/properties that make it an eloquent table-bound model.

File: api/app/Models/Role.php [hunk 3]
```diff
@@ -29,16 +30,21 @@ class Role extends BaseModel
...
-    protected $fillable = ['user_id','role_key'];
+    protected $fillable = ['key','description', 'is_default'];
+
+    protected static $validationRules = [
+        'key' => 'required|rbac_role_exists',
+    ];
+
+    public function permissions()
+    {
+        return new RolePermissionRelation($this->key);
```
Comment (zakhenry) [inline api/app/Models/Role.php:50]: This is a nice and clean pattern. I like it :smile:

**Pull request merged**